File: gui/frontend/src/api/reliabilityProgram.ts
import { api } from './client'

export interface FMEAInput {
  id: string; item: string; function: string; failure_mode: string
  local_effect: string; end_effect: string; cause: string; current_controls: string
  severity: number; occurrence: number; detection: number
  recommended_action: string; action_owner: string; action_status: string
  linked_hazard_ids: string[]; linked_fracas_ids: string[]
  failure_rate?: number; mode_ratio?: number; effect_probability?: number; mission_time?: number
}
export interface HazardInput {
  id: string; title: string; description: string; cause: string
  initial_probability: 'A'|'B'|'C'|'D'|'E'|'F'; initial_severity: 'I'|'II'|'III'|'IV'
  mitigation: string; verification: string
  residual_probability: 'A'|'B'|'C'|'D'|'E'|'F'; residual_severity: 'I'|'II'|'III'|'IV'
  acceptance_status: string; acceptance_authority: string; linked_fmea_ids: string[]
}
export interface FRACASInput {
  id: string; system: string; failure_mode: string; symptom: string
  exposure_at_event?: number; root_cause: string; corrective_action: string
  status: string; action_owner: string; effectiveness_verified: boolean
  recurrence: boolean; downtime: number; linked_fmea_ids: string[]
}
export interface RequirementInput {
  id: string; statement: string; measure: string; target: string; confidence: string
  mission_profile: string; failure_definition: string; verification_method: string
  owner: string; status: string; evidence_ids: string[]
}
export interface TestabilityInput {
  id: string; description: string; weight: number; detected: boolean
  ambiguity_group_size: number; detecting_test_ids: string[]
}
export interface RCMInput {
  id: string; item: string; function: string; functional_failure: string
  failure_mode: string; consequence: string; task_type: string
  task_interval?: number; decision_status: string; rationale: string
  linked_fmea_ids: string[]
}

export type FMEAKind = 'dfmea' | 'pfmea' | 'fmea_msr'
export type ActionPriority = 'H' | 'M' | 'L'
export type FMEAVocabularyDomain =
  'function_verb' | 'failure_deviation' | 'effect_level'
  | 'prevention_control' | 'detection_control'
  | 'verification_method' | 'operating_mode'

export interface FMEAVocabularyTerm {
  id: string
  domain: FMEAVocabularyDomain
  label: string
  category: string
  definition: string
  selection_question: string
  use_when: string
  avoid_when: string
  aliases: string[]
  examples: string[]
  applicability: ('dfmea'|'pfmea'|'fmea_msr'|'all')[]
  source: string
  built_in: boolean
}

export interface FMEAVocabularyAlias {
  id: string
  term_id: string
  value: string
}

export interface FMEAVocabularyProfile {
  version: 1
  custom_terms: FMEAVocabularyTerm[]
  custom_aliases: FMEAVocabularyAlias[]
}

export interface FMEAPlanning {
  company: string; location: string; customer: string; model_program: string
  subject: string; scope: string; exclusions: string; intent: string
  timing: string; tasks: string; tools: string; team: string[]
  owner: string; confidentiality: string; start_date?: string
  revision_date?: string; assumptions: string
  foundation_source_id?: string; foundation_source_revision?: string
  foundation_checksum?: string
}

export interface FMEAStructureSourceRef {
  module: 'prediction'
  analysis_id: string
  analysis_name: string
  entity_type: 'system'|'block'|'part'
  entity_id: string
  parent_entity_id?: string
  imported_at: string
  source_checksum: string
  source_name: string
  /** Stable identity for one quantity-one projection of a grouped source part. */
  piece_key?: string
  reference_designators: string[]
  part_number?: string
  quantity?: number
  manufacturer?: string
  category?: string
}

export interface FMEAStructureNode {
  id: string; name: string; level: string; parent_id?: string
  description: string; interface: string; element_type?: string
  source_ref?: FMEAStructureSourceRef
}

export interface FMEAFunction {
  id: string; structure_node_id: string; description: string
  canonical_verb_id?: string
  function_type: 'primary'|'supporting'|'interface'|'monitoring'|'system_response'
  operating_modes: string[]; owner: string; notes: string
}

export interface FMEAFunctionLink {
  id: string; source_function_id: string; target_function_id: string
  relationship: 'decomposes_to'|'depends_on'|'provides_input'|'enables'|'monitors'|'responds_to'
  label: string; rationale: string
}

export interface FMEAFunctionalRequirement {
  id: string; statement: string
  requirement_type: 'functional'|'performance'|'interface'|'safety'|'regulatory'|'customer'|'process'|'other'
  measure: string; target: string; unit: string; acceptance_criteria: string
  operating_condition: string; source: string; owner: string; confidence: string
  verification_method: string; verification_method_id?: string
  evidence_ids: string[]
  special_characteristic: string
  linked_program_requirement_id?: string; source_checksum?: string
}

export interface FMEAFunctionRequirementLink {
  id: string; function_id: string; requirement_id: string
  strength: 'weak'|'medium'|'strong'; rationale: string
}

export interface FMEABlockDiagramNode {
  id: string
  kind: 'structure'|'external'
  structure_node_id?: string
  /** Diagram-only containment; semantic parentage remains in Structure Analysis. */
  container_parent_block_id?: string
  expanded?: boolean
  label: string
  external_kind?: 'adjacent_system'|'person'|'environment'|'other'
  x: number
  y: number
  width: number
  height: number
  inside_boundary: boolean
}

export interface FMEABlockDiagram {
  version: 2
  density: 'dense'|'compact'|'comfortable'|'spacious'|'expanded'
  boundary: {
    label: string
    x: number
    y: number
    width: number
    height: number
  }
  nodes: FMEABlockDiagramNode[]
  viewport: { x: number; y: number; zoom: number }
  snap_to_grid: boolean
}

export interface FMEAInterface {
  id: string; name: string
  interface_type:
    'physical'|'energy'|'information'|'material'|'human_machine'|'clearance'
  source_block_id?: string; target_block_id?: string
  source_handle?: string; target_handle?: string
  linkage: 'direct'|'indirect'
  directionality: 'directed'|'bidirectional'|'undirected'
  relationship_strength: 'strong'|'weak'|'unspecified'
  relationship_nature: 'beneficial'|'harmful'|'mixed'|'unspecified'
  interface_detail: string
  source_structure_node_id?: string; target_structure_node_id?: string
  external_source: string; external_target: string
  flow_description: string; operating_condition: string
  function_ids: string[]; requirement_ids: string[]
}

export interface FMEAPDiagramItem {
  id: string
  category: 'signal_input'|'intended_output'|'control_factor'|'noise_factor'|'error_state'
  label: string; description: string; requirement_ids: string[]
}

export interface FMEAPDiagram {
  id: string; title: string; primary_function_id: string
  supporting_function_ids: string[]; items: FMEAPDiagramItem[]
}

export interface FMEAAction {
  id: string; kind: 'prevention'|'detection'|'design'|'process'
  description: string; owner: string; target_date?: string
  completion_date?: string
  status: 'open'|'decision_pending'|'implementation_pending'|'completed'|'not_implemented'
  evidence_ids: string[]; decision_rationale: string
}

export interface FMEAFailureChain {
  id: string; function_id?: string
  effect: string
  effect_contexts: {
    id: string; context: string; level_id?: string
    description: string; severity: number
  }[]
  failure_mode: string; deviation_id?: string
  cause: string; effect_level: string; effect_level_id?: string
  severity: number; occurrence?: number; detection?: number
  frequency?: number; monitoring?: number
  prevention_controls: string; prevention_control_method_id?: string
  detection_controls: string; detection_control_method_id?: string
  severity_rationale: string; occurrence_rationale: string
  detection_rationale: string; frequency_rationale: string
  monitoring_rationale: string
  actions: FMEAAction[]; no_action_justification: string
  post_severity?: number; post_occurrence?: number; post_detection?: number
  post_frequency?: number; post_monitoring?: number
  post_mitigated_severity?: number
  post_severity_rationale: string
  linked_hazard_ids: string[]; linked_fracas_ids: string[]
  monitoring_system: string; system_response: string; safe_state: string
  mitigated_effect: string; mitigated_severity?: number
  response_time?: number; fault_tolerant_interval?: number
  management_review_status: string
  management_review_evidence_ids: string[]
  remarks: string
}

export interface FMEAControlPlanRow {
  id: string; failure_chain_id?: string; process_step: string
  product_characteristic: string; process_characteristic: string
  specification: string; measurement_method: string; sample_size: string
  frequency: string; control_method: string; reaction_plan: string
  responsibility: string; special_characteristic: string
  source_revision?: string; stale: boolean
}

export interface FMEARatingCriterion {
  rating: number; label: string; description: string
}

export interface FMEARatingProfile {
  id: string; name: string; version: string; kind: FMEAKind
  built_in: boolean; approved: boolean; approved_by?: string
  approved_date?: string; method_status: string
  rating_axes: Record<string, FMEARatingCriterion[]>
  ap_model: 'aiag_vda_sod_2019'|'aiag_vda_sfm_2019'
  checksum?: string
}

export interface AIAGVDAFMEAAnalysis {
  id: string; name: string; kind: FMEAKind; revision: string
  status: 'draft'|'in_review'|'finalized'
  rating_profile_id?: string; planning: FMEAPlanning
  structure_nodes: FMEAStructureNode[]; functions: FMEAFunction[]
  block_diagram: FMEABlockDiagram
  function_links: FMEAFunctionLink[]
  functional_requirements: FMEAFunctionalRequirement[]
  function_requirement_links: FMEAFunctionRequirementLink[]
  interfaces: FMEAInterface[]
  p_diagrams: FMEAPDiagram[]
  failure_chains: FMEAFailureChain[]; control_plan: FMEAControlPlanRow[]
  parent_dfmea_id?: string; source_revision?: string
  standalone_justification: string
  template_source_id?: string; template_source_revision?: string
  template_source_checksum?: string
}

export interface FMEAIssue {
  step: number; code: string; severity: 'error'|'warning'
  record_id?: string; field?: string; message: string; analysis_id?: string
}

export interface AIAGVDAFMEAResult extends AIAGVDAFMEAAnalysis {
  failure_chains: (FMEAFailureChain & {
    action_priority: ActionPriority; post_action_priority: ActionPriority|null
    action_priority_meaning: string; action_priority_severity: number
  })[]
  rating_profile: Pick<FMEARatingProfile, 'id'|'name'|'version'|'checksum'|'method_status'>
  issues: FMEAIssue[]
  step_readiness: { step: number; ready: boolean; errors: number; warnings: number }[]
  finalization_ready: boolean
  control_plan_review: {
    failure_chain_id: string; status: 'missing'|'different'|'in_sync'
    differences: { field: string; current: string; proposed: string }[]
    proposal: FMEAControlPlanRow
  }[]
  function_analysis_summary: {
    functions: number; primary_functions: number; requirements: number
    correlations: number; interfaces: number; p_diagrams: number
    structures_with_functions: number; structures_total: number
    functions_with_requirements: number; functions_with_failure_chains: number
    stale_requirement_links: number; coverage_gaps: number
  }
  function_coverage: {
    structure_node_id: string; structure_name: string; level: string
    function_ids: string[]; requirement_ids: string[]
    interface_ids: string[]; failure_chain_ids: string[]; gaps: string[]
  }[]
  requirement_sync: {
    requirement_id: string; source_id: string|null
    status: 'local'|'in_sync'|'stale'|'missing_source'|'unbaselined'
    stored_checksum: string|null; current_checksum: string|null
    differences: { field: string; snapshot: unknown; source: unknown }[]
  }[]
  summary: {
    failure_chains: number; high_action_priority: number
    medium_action_priority: number; low_action_priority: number
    post_high_action_priority: number; open_actions: number
    overdue_actions: number; errors: number; warnings: number
  }
  methodology: {
    title: string; edition: string; errata: string
    implementation_status: string; method_version: string
    profile_checksum?: string; rpn_calculated: false; interpretation: string
  }
}

export interface ReliabilityProgramRequest {
  fmea: FMEAInput[]; hazards: HazardInput[]; fracas: FRACASInput[]
  requirements: RequirementInput[]; testability_faults: TestabilityInput[]; rcm: RCMInput[]
  fmea_analyses: AIAGVDAFMEAAnalysis[]; rating_profiles: FMEARatingProfile[]
  total_exposure?: number; CI: number; isolation_threshold: number
  medium_rpn: number; high_rpn: number
}

export interface ReliabilityProgramResponse {
  fmea: {
    rows: (FMEAInput & { rpn: number; screening_band: string; mode_criticality: number | null })[]
    ranked_ids: string[]
    summary: { total: number; open_actions: number; high_or_severity_override: number; criticality_available: number; total_mode_criticality: number }
    rpn_policy: { method: string; medium_threshold: number; high_threshold: number; severity_override: string; warning: string }
  }
  aiag_vda_fmea: {
    analyses: AIAGVDAFMEAResult[]
    summary: {
      analyses: number; dfmea: number; pfmea: number; fmea_msr: number
      high_action_priority: number; open_actions: number
      finalization_ready: number; issues: number
    }
    issues: FMEAIssue[]; rating_profiles: FMEARatingProfile[]
    methodology: {
      title: string; edition: string; errata: string
      implementation_status: string; method_version: string
    }
  }
  hazards: {
    rows: (HazardInput & { initial_risk: RiskCell; residual_risk: RiskCell; risk_reduced: boolean; risk_worsened: boolean })[]
    summary: { total: number; initial_high_or_serious: number; residual_high_or_serious: number; unaccepted: number; worsened: number }
    method: string; warning: string
  }
  fracas: {
    summary: { records: number; open: number; closed: number; effectiveness_verified: number; recurrences: number; closure_fraction: number | null; verification_fraction: number | null; total_downtime: number }
    exposure_metrics: { total_exposure: number | null; event_rate: number | null; mtbf: number | null; rate_lower: number | null; rate_upper: number | null; confidence_level: number }
    pareto_failure_modes: { name: string; count: number }[]
    pareto_systems: { name: string; count: number }[]
    warning: string
  }
  requirements: { rows: { id: string; missing_fields: string[]; evidence_count: number; status: string; verification_ready: boolean }[]; summary: { total: number; complete_definitions: number; with_evidence: number; verification_ready: number }; warning: string }
  testability: { rows: (TestabilityInput & { isolation_eligible: boolean })[]; summary: { faults: number; total_weight: number; fraction_faults_detected: number; fraction_faults_isolated: number; isolation_threshold: number; undetected_fault_ids: string[] }; method: string; warning: string } | null
  rcm: { summary: { items: number; unresolved: number; with_interval: number }; consequences: Record<string, number>; tasks: Record<string, number>; unresolved_ids: string[]; warning: string }
  traceability: {
    summary: { links: number; resolved_links: number; unknown_references: number; missing_reciprocal_links: number; issues: number }
    issues: { code: string; source_id: string; field: string; target_id: string; expected_record_type?: string; expected_reciprocal_field?: string }[]
    warning: string
  }
  standards_context: { status: string; references: string[] }
}

interface RiskCell { probability: string; severity: string; risk_index: number; risk_level: string }

export const analyzeReliabilityProgram = (request: ReliabilityProgramRequest) =>
  api.post<ReliabilityProgramResponse>('/reliability-program/analyze', request).then(response => response.data)

export const getFmeaRatingProfiles = () =>
  api.get<FMEARatingProfile[]>('/reliability-program/fmea/rating-profiles')
    .then(response => response.data)

File: src/reliability/Reliability_program.py
"""Reliability-program workflow calculations and integrity checks.

The objects here support FMEA/FMECA, MIL-STD-882E-style hazard risk,
FRACAS, requirement traceability, testability, and RCM records. Ordinal
products such as RPN are explicitly screening aids, not calibrated risk.
"""

from __future__ import annotations

from collections import Counter
import math
from typing import Iterable

from scipy.stats import chi2

from reliability.AIAG_VDA_FMEA import analyze_aiag_vda_fmea


# MIL-STD-882E Table III risk assessment matrix. Tuple is (risk index, level).
_HAZARD_MATRIX = {
    "A": {"I": (1, "high"), "II": (3, "high"), "III": (7, "serious"), "IV": (13, "medium")},
    "B": {"I": (2, "high"), "II": (5, "high"), "III": (9, "serious"), "IV": (16, "medium")},
    "C": {"I": (4, "high"), "II": (6, "serious"), "III": (11, "medium"), "IV": (18, "low")},
    "D": {"I": (8, "serious"), "II": (10, "medium"), "III": (14, "medium"), "IV": (19, "low")},
    "E": {"I": (12, "medium"), "II": (15, "medium"), "III": (17, "low"), "IV": (20, "low")},
    "F": {"I": (0, "eliminated"), "II": (0, "eliminated"), "III": (0, "eliminated"), "IV": (0, "eliminated")},
}


def _risk(probability: str, severity: str) -> dict:
    probability = str(probability).upper()
    severity = str(severity).upper()
    if probability not in _HAZARD_MATRIX or severity not in _HAZARD_MATRIX[probability]:
        raise ValueError("Hazard probability must be A–F and severity must be I–IV.")
    index, level = _HAZARD_MATRIX[probability][severity]
    return {"probability": probability, "severity": severity,
            "risk_index": index, "risk_level": level}


def analyze_fmea(rows: Iterable[dict], *, medium_rpn: int = 100,
                 high_rpn: int = 200) -> dict:
    if not 0 < medium_rpn < high_rpn:
        raise ValueError("FMEA RPN thresholds must satisfy 0 < medium < high.")
    output = []
    for raw in rows:
        severity = int(raw["severity"])
        occurrence = int(raw["occurrence"])
        detection = int(raw["detection"])
        if any(not 1 <= value <= 10 for value in (severity, occurrence, detection)):
            raise ValueError("FMEA severity, occurrence, and detection must be integers from 1 to 10.")
        rpn = severity * occurrence * detection
        band = "high" if rpn >= high_rpn else "medium" if rpn >= medium_rpn else "low"
        if severity >= 9 and band != "high":
            band = "severity_override"
        rate = raw.get("failure_rate")
        ratio = raw.get("mode_ratio")
        effect = raw.get("effect_probability")
        mission = raw.get("mission_time")
        criticality = None
        if any(value is not None for value in (rate, ratio, effect, mission)):
            if any(value is None for value in (rate, ratio, effect, mission)):
                raise ValueError(
                    "FMECA criticality requires failure_rate, mode_ratio, "
                    "effect_probability, and mission_time together.")
            rate, ratio, effect, mission = map(float, (rate, ratio, effect, mission))
            if (not all(math.isfinite(value) for value in (rate, ratio, effect, mission))
                    or rate < 0 or not 0 <= ratio <= 1
                    or not 0 <= effect <= 1 or mission <= 0):
                raise ValueError(
                    "FMECA criticality requires finite rate >= 0, mode ratio and "
                    "effect probability in [0, 1], and mission time > 0.")
            criticality = rate * ratio * effect * mission
        output.append({**raw, "rpn": rpn, "screening_band": band,
                       "mode_criticality": criticality})
    ordered = sorted(output, key=lambda row: (-row["severity"], -row["rpn"], str(row.get("id", ""))))
    return {
        "rows": output,
        "ranked_ids": [row.get("id") for row in ordered],
        "summary": {
            "total": len(output),
            "open_actions": sum(str(row.get("action_status", "open")).lower() not in {"closed", "verified"} for row in output),
            "high_or_severity_override": sum(row["screening_band"] in {"high", "severity_override"} for row in output),
            "criticality_available": sum(row["mode_criticality"] is not None for row in output),
            "total_mode_criticality": sum(
                row["mode_criticality"] or 0.0 for row in output),
        },
        "rpn_policy": {
            "method": "ordinal_product_screening",
            "medium_threshold": medium_rpn, "high_threshold": high_rpn,
            "severity_override": "severity >= 9",
            "warning": "RPN ranks records from ordinal inputs; use consequence and evidence fields for risk decisions.",
        },
    }


def analyze_hazards(rows: Iterable[dict]) -> dict:
    output = []
    for raw in rows:
        initial = _risk(raw["initial_probability"], raw["initial_severity"])
        residual = _risk(raw["residual_probability"], raw["residual_severity"])
        worsened = (residual["risk_index"] != 0 and initial["risk_index"] != 0
                    and residual["risk_index"] < initial["risk_index"])
        output.append({**raw, "initial_risk": initial, "residual_risk": residual,
                       "risk_reduced": residual["risk_index"] == 0
                       or (initial["risk_index"] != 0 and residual["risk_index"] > initial["risk_index"]),
                       "risk_worsened": worsened})
    return {
        "rows": output,
        "summary": {
            "total": len(output),
            "initial_high_or_serious": sum(row["initial_risk"]["risk_level"] in {"high", "serious"} for row in output),
            "residual_high_or_serious": sum(row["residual_risk"]["risk_level"] in {"high", "serious"} for row in output),
            "unaccepted": sum(str(row.get("acceptance_status", "pending")).lower() not in {"accepted", "closed"} for row in output),
            "worsened": sum(row["risk_worsened"] for row in output),
        },
        "method": "MIL-STD-882E Table III risk assessment matrix",
        "warning": "Risk index is an ordinal matrix category. Record the supporting evidence and acceptance authority with the decision.",
    }


def analyze_fracas(rows: Iterable[dict], *, total_exposure: float | None = None,
                   CI: float = 0.95) -> dict:
    records = list(rows)
    if not 0 < CI < 1:
        raise ValueError("FRACAS confidence must be strictly between 0 and 1.")
    count = len(records)
    open_count = sum(str(row.get("status", "open")).lower() not in {"closed", "verified"} for row in records)
    verified = sum(bool(row.get("effectiveness_verified", False)) for row in records)
    recurrence = sum(bool(row.get("recurrence", False)) for row in records)
    modes = Counter(str(row.get("failure_mode", "Unspecified")).strip() or "Unspecified" for row in records)
    systems = Counter(str(row.get("system", "Unspecified")).strip() or "Unspecified" for row in records)
    rate = mtbf = lower = upper = None
    if total_exposure is not None:
        exposure = float(total_exposure)
        if not math.isfinite(exposure) or exposure <= 0:
            raise ValueError("FRACAS total_exposure must be positive and finite.")
        rate = count / exposure
        mtbf = exposure / count if count else None
        tail = (1.0 - CI) / 2.0
        lower = 0.0 if count == 0 else float(0.5 * chi2.ppf(tail, 2 * count) / exposure)
        upper = float(0.5 * chi2.ppf(1.0 - tail, 2 * (count + 1)) / exposure)
    return {
        "summary": {
            "records": count, "open": open_count, "closed": count - open_count,
            "effectiveness_verified": verified, "recurrences": recurrence,
            "closure_fraction": (count - open_count) / count if count else None,
            "verification_fraction": verified / count if count else None,
            "total_downtime": sum(float(row.get("downtime", 0) or 0) for row in records),
        },
        "exposure_metrics": {
            "total_exposure": total_exposure, "event_rate": rate, "mtbf": mtbf,
            "rate_lower": lower, "rate_upper": upper, "confidence_level": CI,
        },
        "pareto_failure_modes": [{"name": key, "count": value} for key, value in modes.most_common()],
        "pareto_systems": [{"name": key, "count": value} for key, value in systems.most_common()],
        "warning": "FRACAS rate uses the supplied aggregate exposure. Interpret changes alongside reporting and detection practices.",
    }


def analyze_requirements(rows: Iterable[dict]) -> dict:
    records = list(rows)
    required = (
        "id", "statement", "measure", "target", "confidence",
        "mission_profile", "failure_definition", "verification_method", "owner",
    )
    detail = []
    for row in records:
        missing = [field for field in required if str(row.get(field, "")).strip() == ""]
        evidence = row.get("evidence_ids", []) or []
        status = str(row.get("status", "draft")).lower()
        detail.append({"id": row.get("id"), "missing_fields": missing,
                       "evidence_count": len(evidence), "status": status,
                       "verification_ready": not missing and bool(evidence)
                       and status in {"verified", "accepted"}})
    return {
        "rows": detail,
        "summary": {
            "total": len(records),
            "complete_definitions": sum(not row["missing_fields"] for row in detail),
            "with_evidence": sum(row["evidence_count"] > 0 for row in detail),
            "verification_ready": sum(row["verification_ready"] for row in detail),
        },
        "warning": "Traceability measures record completeness; assess evidence quality and acceptance in the linked review.",
    }


def analyze_testability(faults: Iterable[dict], *, isolation_threshold: int = 1) -> dict:
    records = list(faults)
    if isolation_threshold < 1:
        raise ValueError("isolation_threshold must be at least 1.")
    weights = [float(row.get("weight", 0)) for row in records]
    ambiguities = [int(row.get("ambiguity_group_size", 0)) for row in records]
    total = sum(weights)
    if (total <= 0 or any(not math.isfinite(value) or value < 0 for value in weights)
            or any(value < 1 for value in ambiguities)):
        raise ValueError(
            "Testability faults require finite non-negative weights with a "
            "positive total and ambiguity-group sizes of at least 1.")
    detected = sum(float(row["weight"]) for row in records if bool(row.get("detected", False)))
    isolated = sum(float(row["weight"]) for row in records
                   if bool(row.get("detected", False))
                   and int(row.get("ambiguity_group_size", 0)) <= isolation_threshold)
    rows = [{**row, "isolation_eligible": bool(row.get("detected", False))
             and int(row.get("ambiguity_group_size", 0)) <= isolation_threshold}
            for row in records]
    return {
        "rows": rows,
        "summary": {
            "faults": len(records), "total_weight": total,
            "fraction_faults_detected": detected / total,
            "fraction_faults_isolated": isolated / total,
            "isolation_threshold": isolation_threshold,
            "undetected_fault_ids": [row.get("id") for row in records if not bool(row.get("detected", False))],
        },
        "method": "weighted FFD/FFI fault-universe accounting",
        "warning": "Coverage reflects the declared fault universe, weights, diagnostic model, and ambiguity groups.",
    }


def analyze_rcm(rows: Iterable[dict]) -> dict:
    records = list(rows)
    consequence = Counter(str(row.get("consequence", "unclassified")).lower() for row in records)
    task = Counter(str(row.get("task_type", "undecided")).lower() for row in records)
    unresolved = [row.get("id") for row in records
                  if str(row.get("decision_status", "open")).lower() not in {"approved", "closed"}]
    return {
        "summary": {"items": len(records), "unresolved": len(unresolved),
                    "with_interval": sum(row.get("task_interval") not in {None, ""} for row in records)},
        "consequences": dict(consequence), "tasks": dict(task),
        "unresolved_ids": unresolved,
        "warning": "Confirm each RCM task against the asset context, failure behavior, and effectiveness evidence.",
    }


def analyze_traceability(*, fmea: Iterable[dict], hazards: Iterable[dict],
                         fracas: Iterable[dict], rcm: Iterable[dict]) -> dict:
    """Check typed links without assuming that an unlinked record is invalid."""
    fmea, hazards, fracas, rcm = map(list, (fmea, hazards, fracas, rcm))
    groups = {
        "fmea": {str(row.get("id")) for row in fmea},
        "hazards": {str(row.get("id")) for row in hazards},
        "fracas": {str(row.get("id")) for row in fracas},
        "rcm": {str(row.get("id")) for row in rcm},
    }
    records = {
        "fmea": list(fmea), "hazards": list(hazards),
        "fracas": list(fracas), "rcm": list(rcm),
    }
    specifications = (
        ("fmea", "linked_hazard_ids", "hazards"),
        ("fmea", "linked_fracas_ids", "fracas"),
        ("hazards", "linked_fmea_ids", "fmea"),
        ("fracas", "linked_fmea_ids", "fmea"),
        ("rcm", "linked_fmea_ids", "fmea"),
    )
    issues = []
    total_links = resolved_links = 0
    for source_group, field, target_group in specifications:
        for row in records[source_group]:
            for target_id in row.get(field, []) or []:
                total_links += 1
                target_id = str(target_id)
                if target_id in groups[target_group]:
                    resolved_links += 1
                else:
                    issues.append({
                        "code": "unknown_reference", "source_id": row.get("id"),
                        "field": field, "target_id": target_id,
                        "expected_record_type": target_group,
                    })
    reciprocal_pairs = (
        ("fmea", "linked_hazard_ids", "hazards", "linked_fmea_ids"),
        ("fmea", "linked_fracas_ids", "fracas", "linked_fmea_ids"),
    )
    indexes = {name: {str(row.get("id")): row for row in rows}
               for name, rows in records.items()}
    for source_group, source_field, target_group, target_field in reciprocal_pairs:
        for source in records[source_group]:
            for target_id in source.get(source_field, []) or []:
                target = indexes[target_group].get(str(target_id))
                if target is not None and str(source.get("id")) not in {
                        str(value) for value in target.get(target_field, []) or []}:
                    issues.append({
                        "code": "missing_reciprocal_link",
                        "source_id": source.get("id"), "field": source_field,
                        "target_id": str(target_id),
                        "expected_reciprocal_field": target_field,
                    })
    unknown = sum(issue["code"] == "unknown_reference" for issue in issues)
    reciprocal = sum(issue["code"] == "missing_reciprocal_link" for issue in issues)
    return {
        "summary": {
            "links": total_links, "resolved_links": resolved_links,
            "unknown_references": unknown,
            "missing_reciprocal_links": reciprocal,
            "issues": len(issues),
        },
        "issues": issues,
        "warning": "Link integrity covers identifier resolution and reciprocal links; review the technical consistency of linked records together.",
    }


def analyze_reliability_program(*, fmea=(), hazards=(), fracas=(), requirements=(),
                                testability_faults=(), rcm=(), total_exposure=None,
                                CI=0.95, isolation_threshold=1,
                                medium_rpn=100, high_rpn=200,
                                fmea_analyses=(), rating_profiles=()) -> dict:
    fmea, hazards, fracas = list(fmea), list(hazards), list(fracas)
    requirements, testability_faults, rcm = (
        list(requirements), list(testability_faults), list(rcm))
    return {
        "fmea": analyze_fmea(fmea, medium_rpn=medium_rpn, high_rpn=high_rpn),
        "aiag_vda_fmea": analyze_aiag_vda_fmea(
            fmea_analyses, rating_profiles=rating_profiles,
            program_requirements=requirements),
        "hazards": analyze_hazards(hazards),
        "fracas": analyze_fracas(fracas, total_exposure=total_exposure, CI=CI),
        "requirements": analyze_requirements(requirements),
        "testability": analyze_testability(testability_faults, isolation_threshold=isolation_threshold)
        if testability_faults else None,
        "rcm": analyze_rcm(rcm),
        "traceability": analyze_traceability(
            fmea=fmea, hazards=hazards, fracas=fracas, rcm=rcm),
        "standards_context": {
            "status": "standards_informed_workflow",
            "references": ["AIAG & VDA FMEA Handbook, First Edition (2019)",
                           "AIAG & VDA FMEA Handbook English Errata v2 (2020)",
                           "IEC 60812:2018 (full-text verification pending)",
                           "MIL-STD-882E Change 1", "MIL-HDBK-338B",
                           "RL-TR-91-180"],
        },
    }


__all__ = [
    "analyze_fmea", "analyze_hazards", "analyze_fracas",
    "analyze_requirements", "analyze_testability", "analyze_rcm",
    "analyze_traceability",
    "analyze_reliability_program",
]
